crates/net/downloaders/src/test_utils/file_client.rs

```diff
@@ -62,19 +69,82 @@ impl FileClient {
     /// Create a new file client from a file path.
     pub async fn new<P: AsRef<Path>>(path: P) -> Result<Self, FileClientError> {
         let file = File::open(path).await?;
-        FileClient::from_file(file)
+        FileClient::from_file(file).await
     }
 
     /// Initialize the [`FileClient`](FileClient) with a file directly.
-    pub(crate) fn from_file(file: File) -> Result<Self, FileClientError> {
-        let reader = BufReader::new(file);
-
-        Ok(Self {
-            reader,
-            headers: HashMap::new(),
-            bodies: HashMap::new(),
-            is_syncing: Arc::new(Default::default()),
-        })
+    pub(crate) async fn from_file(file: File) -> Result<Self, FileClientError> {
```

Comment:
I don't know anything about the file layout or how the data in it is encoded.

Reply:
It is just RLP encoded blocks one after the other, it is not encoded as a list so we do not know the size prior to reading :/

Since we download in reverse, I think we need to load the entire file to memory, which is kind of annoying. An alternative is to decode each block and save a list of offsets to seek to later, i.e.:

1. Save current offset as `block -> offset`
2. Decode block
3. Go to 1.

And then keep the last `n` blocks in memory so we don't need to decode them again (since these are imported first anyway).

A final alternative is to support downloading in a forward fashion (which is also needed for e.g. Clique consensus) and just use that - then we can decode blocks lazily.